File: winekloo/frontend/lib/repositories/restaurateur_repo.dart
import 'dart:async';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/restaurateur_model.dart';

class RestaurateurRepository {
  final database = Supabase.instance.client.from('restaurateur');

  Future createRestaurateur(Restaurateur newRestaurateur) async {
    final response = await database.insert({
      'DiningType': newRestaurateur.diningType,
      'Name': newRestaurateur.name,
      'Email': newRestaurateur.email,
      'PhoneNumber': newRestaurateur.phoneNumber,
      'Password': newRestaurateur.password,
      'Location': newRestaurateur.location,
      'Photo': newRestaurateur.photo,
      'WorkingHours': newRestaurateur.workingHours,
      'WorkingDays': newRestaurateur.workingDays,
      'Description': newRestaurateur.description,
      'RatingValueAverage': newRestaurateur.ratingValueAverage?.toString(),
      'MenuFilePDF': newRestaurateur.menuFilePDF,
      'Categories': newRestaurateur.restaurateurCategories,
      'DietaryPreferences': newRestaurateur.restaurateurPreferences,
      'SpecialFeatures': newRestaurateur.restaurateurSpecialFeatures,
      'Pricing': newRestaurateur.restaurateurPricing,
    });
    return response;
  }

  Future<List<Restaurateur>> getRestaurateurs() async {
    final response = await database.select();
    return (response as List)
        .map((data) => Restaurateur.fromJson(data))
        .toList();
  }

  Future<Restaurateur?> getOneRestaurateur(int restaurateurID) async {
    try {
      final response = await database.select().eq('RestaurateurID', restaurateurID).single();
      return Restaurateur.fromJson(response);
    } catch (e) {
      print('Error fetching restaurateur: $e');
      return null;  
    }
  }

  Future<void> updateRestaurateur(Restaurateur updatedRestaurateur) async {
    final response = await database
      .update(updatedRestaurateur.toJson())
      .eq('RestaurateurID', updatedRestaurateur.restaurateurID!);

    if (response.error != null) {
      throw Exception('Failed to update restaurateur: ${response.error!.message}');
    }
  }

  Future deleteRestaurateur(int restaurateurID) async {
    final response = await database.delete().eq('RestaurateurID', restaurateurID);
    if (response.error != null) {
      print('Error deleting restaurateur: ${response.error!.message}');
      return;
    }
    print('Successfully deleted restaurateur with ID: $restaurateurID');
  }

  final stream = Supabase.instance.client.from('restaurateur').stream(
    primaryKey: ['restaurateurID'],
  );
}

File: winekloo/backend/app/repositories/restaurateur_repo.py
from app.db import db

class RestaurateurRepository:
    
    @staticmethod
    def create_restaurateur(new_restaurateur):
        response = db.table('restaurateur').insert({
            'DiningType': new_restaurateur['diningType'],
            'Name': new_restaurateur['name'],
            'Email': new_restaurateur['email'],
            'PhoneNumber': new_restaurateur['phoneNumber'],
            'Password': new_restaurateur['password'],
            'Location': new_restaurateur['location'],
            'Photo': new_restaurateur['photo'],
            'WorkingHours': new_restaurateur['workingHours'],
            'WorkingDays': new_restaurateur['workingDays'],
            'Description': new_restaurateur['description'],
            'RatingValueAverage': new_restaurateur['ratingValueAverage'],
            'MenuFilePDF': new_restaurateur['menuFilePDF'],
            'Categories': new_restaurateur['restaurateurCategories'],
            'DietaryPreferences': new_restaurateur['restaurateurPreferences'],
            'SpecialFeatures': new_restaurateur['restaurateurSpecialFeatures'],
            'Pricing': new_restaurateur['restaurateurPricing']
        }).execute()

        return response.data if response.data else None

    @staticmethod
    def get_restaurateurs():
        response = db.table('restaurateur').select('*').execute()
        return response.data if response.data else None

    @staticmethod
    def get_one_restaurateur(restaurateur_id):
        response = db.table('restaurateur').select('*').eq('RestaurateurID', restaurateur_id).single().execute()
        return response.data if response.data else None

    @staticmethod
    def update_restaurateur(updated_restaurateur):
        response = db.table('restaurateur').update(updated_restaurateur).eq('RestaurateurID', updated_restaurateur['restaurateurID']).execute()
        return response.data if response.data else None

    @staticmethod
    def delete_restaurateur(restaurateur_id):
        response = db.table('restaurateur').delete().eq('RestaurateurID', restaurateur_id).execute()
        return response.data if response.data else None
